lib/actions/preload.coffee

```diff
@@ -0,0 +1,206 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+###
+
+_ = require('lodash')
+form = require('resin-cli-form')
+preload = require('resin-preload')
+Promise = require('bluebird')
+resin = require('resin-sdk-preconfigured')
+
+dockerUtils = require('../utils/docker')
+
+LATEST = 'latest'
+
+buildExpandOptions =
+	build:
+		$select: ['id', 'commit_hash', 'push_timestamp']
+		$orderby: 'push_timestamp desc'
+		$filter:
+			status: 'success'
+
+getApplicationsWithSuccessfulBuilds = (resin, deviceType) ->
+	resin.pine.get
+		resource: 'my_application'
+		options:
+			filter:
+				device_type: deviceType
+				build:
+					$any:
+						$alias: 'b'
+						$expr:
+							b:
+								status: 'success'
+			expand:
+				build:
+					$select: [ 'id', 'commit_hash', 'push_timestamp', 'status' ]
+					$orderby: 'push_timestamp desc'
+					# the filter is commented because it causes an api error
+					#$filter:
+					#	status: 'success'
+			select: [ 'id', 'app_name', 'device_type' ]
+			orderby: 'app_name asc'
+	# manual filtering
+	.then (applications) ->
+		applications.forEach (application) ->
+			application.build = application.build.filter (build) ->
+				build.status == 'success'
+		applications
+
+selectApplication = (resin, form, deviceType) ->
+	getApplicationsWithSuccessfulBuilds(resin, deviceType)
+	.then (applications) ->
+		if applications.length == 0
+			expectedError("You have no apps for a '#{deviceType}' device type.")
+		form.ask
+			message: 'Select an application'
+			type: 'list'
+			choices: applications.map (app) ->
+				name: app.app_name
+				value: app
+
+selectApplicationCommit = (resin, form, builds) ->
+	if builds.length == 0
+		expectedError('This application has no successful builds.')
+	DEFAULT_CHOICE = {'name': LATEST, 'value': LATEST}
+	choices = [ DEFAULT_CHOICE ].concat builds.map (build) ->
+		name: "#{build.push_timestamp} - #{build.commit_hash}"
+		value: build.commit_hash
+	return form.ask
+		message: 'Select a build'
+		type: 'list'
+		default: LATEST
+		choices: choices
+
+disableAutomaticUpdates = (form, resin, application, commit) ->
+	if commit == LATEST or not application.should_track_latest_release
+		return Promise.resolve()
+	message = '''
+
+		This application is set to automatically update all devices to the latest
+		available version.
+		You need to disable this if you want to preload a specific build.
+		Do you want to disable automatic updates for this application?
+	'''
+	form.ask
+		message: message,
+		type: 'confirm'
+	.then (update) ->
+		if not update
+			return
```

Comment:
Based on this, it sounds like preloading + automatic updates isn't useless, and makes sense in lots of environments where you might use a device first time without reliable internet, or where you want to start using the device immediately while letting the image download slowly start up in the background (and later upgrade with a handover or something). Is that right?

If so, sounds to me like we should tone this message down further, since that seems like a pretty reasonable use case. I would have thought lots (most?) preloaded images really are being used to stop devices _requiring_ internet on first boot, not to avoid automatic deploys entirely. Could we tweak the message a bit further to something like:

```
This application is set to automatically update all devices to the latest available version.
This might be unexpected behaviour: with this enabled, the preloaded device will still
download and install the latest build once it is online.

Do you want to disable automatic updates for this application?
```

And then drop the warning, since it's not the scary result I thought it was.

Are my assumptions right, and does that make sense do you think?